src/os/bluestore/BlueFS.cc

```diff
@@ -1228,9 +1233,7 @@ void BlueFS::_compact_log_async(std::unique_lock<std::mutex>& l)
   dout(10) << __func__ << " new_log_jump_to 0x" << std::hex << new_log_jump_to
 	   << std::dec << dendl;
 
-  // create a new log [writer]
-  new_log = new File;
-  new_log->fnode.ino = 0;   // so that _flush_range won't try to log the fnode
+  // allocate
```

Comment:
@liewegas But we haven't dropped the **lock** until we call `flush_bdev()` later. How is this supposed to cause a race condition?

Reply:
That's exactly where the lock is dropped and the other thread starts its own compaction attempt.  Moving this even further up to the top, right after the assert, just so that it is obvious it's being used as the "in progress" flag.

Reply:
> That's exactly where the lock is dropped and the other thread starts its own compaction attempt

@liewegas Sorry, I am still not convinced by the above explanation.

Thread A: call `sync_metadata` -> acquire `lock  `-> check **new_log** (suppose it is null) -> call `_compact_log_async` and set **new_log** -> drop `lock` -> call `flush_bdev` ...
Thread B: call `sync_metadata` -> try acquire `lock`  but blocked by Thread A -> successfully acquire `lock ` -> check  **new_log** (it is already set by Thread A) -> Thread B knows now the compaction is in progress and quit

Since we are using a execlusive mutex to blockingly set and clear **new_log**, why moving this up to the top(both are under the protection of the `lock`) shall make such a difference?